### ISSUE TITLE
Slight difference between Deconvnet and Guided BP

### ISSUE THREAD

**Issue by yutongy98** [kazuto1011/grad-cam-pytorch#34]
Thanks for the GREAT repo! I noticed, there is only one small difference between the algorithms of this 2 visualization methods:

For Guided BP we use: **return (F.relu(grad_in[0]),)**
For Deconvnet we use: **return (F.relu(grad_out[0]),)**

For me the Guided BP is understandable, but I am confused about the deconvnet. The deconvnet consist of  _unpooling_, _ReLu_  and _deconvolution_ layers(https://www.quora.com/How-does-a-deconvolutional-neural-network-work), but I only find the _ReLu_ operation is implemented by using F.relu. Maybe I wrongly understood the Deconvnet visulization method or miss something about the use of Pytorch. I hope you can give me the right direction!

Thanks a lot.

**Comment by kazuto1011** [comment]
> The deconvnet consist of unpooling, ReLu and deconvolution layers.

The unpooling and deconvolution are the backward routing of pooling and convolution, respectively. The relu is a negative clipping of gradient flows. Therefore you can say the guided bp consists of unpooling, relu, deconvolution, and **backward activation**. The difference is just the backward activation, which routes gradients based on the forward pass, not the gradients themselves. The relevant papers only consider relu activation, _i.e._ backward relu.

||deconvolution|backward relu|gradient relu|unpooling|
|:-:|:-:|:-:|:-:|:-:|
|vanilla bp|&#10003;|&#10003;||&#10003;|
|deconvnet|&#10003;||&#10003;|&#10003;|
|guided bp|&#10003;|&#10003;|&#10003;|&#10003;|

> For Guided BP we use: return (F.relu(grad_in[0]),)
For Deconvnet we use: return (F.relu(grad_out[0]),)

`grad_in` is a gradient after the backward relu, while `grad_out` is a gradient before the backward relu. `F.relu()` is the gradient relu. In vanilla backpropagation, we return the raw `grad_in` to the next layer.

**Comment by yutongy98** [comment]
Thanks for the helpful and fast reply! 
So in conclusion, the sole difference between the 3 approaches, is how they backpropagate through the **ReLu**(As show in the above table). As for the rest parts, i. e. _unpooling_ layers and _deconvolution_ layers, they are used in the level of algorithm implementation with no difference. We just focus on the _ReLu_ part in the backward route when we want to get the different kinds of saliency map. Am I right?

**Comment by kazuto1011** [comment]
Yes. Figure 1 of the guided bp paper is helpful to understand this point (https://arxiv.org/pdf/1412.6806.pdf).

**Comment by yutongy98** [comment]
Thanks again for you kind help! Now this 3 methods are clearer to me : )
